Move GeoAxes.background_patch to GeoAxes.patch.

QuLogic opened the pull request:
## Rationale

This is a followup to #1213, converting the Cartopy-specific `background_patch` into the usual `Axes.patch` from Matplotlib.

## Implications

Unfortunately, this breaks several tests. Most of these are due to clipping bugs upstream in Matplotlib (at least matplotlib/matplotlib#15952). I'm hoping to get these fixed, regenerate any test images with the new version if necessary, then just increase tolerance for the old images (_most_ of them are small changes anyway.)

Since we aren't using the default patch, this should fix tight layout too, but I haven't yet tested it.

Fixes #1207.

dopplershift commented:
In general this looks good. Everything seems cleaner. Waiting to see what the image changes look like..

QuLogic commented:
Currently, it's mostly the same as the referenced Matplotlib issue, plus matplotlib/matplotlib#15946.

greglucas commented:
Looks great to me. I just pulled your code and this does appear to fix the tight layout issues. Specifically, #1207 looks good to me locally with your changes.

With constrained_layout it appears much better, but there is still a slight shift between the two figures.

```
import cartopy
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

f, ax1 = plt.subplots(1, 1, constrained_layout=True,
                      subplot_kw=dict(projection=ccrs.PlateCarree()))

f.savefig('mpl-1.png')
f.savefig('mpl-2.png')
```

QuLogic commented:
For reference, I've committed the changed images and the diffs [here](https://github.com/QuLogic/cartopy/commit/67da84c0841b49f190983b7bf3d4bd173b88dc18). To understand most of the changes, you have to know that Matplotlib draws spines at 0.8 linewidth, so there's usually one middle pixel full black, and two surrounding pixels of grey. Correctly clipped artists should end at the inner grey or the black pixel, and the inner grey one might be shaded a bit by the artist's colour. If not clipped correctly, then they might mix into the outer grey pixels, or only mix into the inner grey one via antialiasing (and thus a bit darker than they should be.)

My analysis of the changes is:
* `contour_label.png` 0.9719825101307121. The difference is along the left/right edges, but honestly can't see it.
* `gshhs_coastlines.png` 0.006666666666666667. Most tests use 0.5 tolerance, so I just increased that. It seems like some of the coastlines near the edge changed, so I'm not exactly sure why, but it could be because of the way it clipped.
* `regrid_image.png` 5.5476486310267825. Mainly the clipping now works for the first Axes. However, there is an annoying extra bit of clipped pixels on the red/green/blue images, mostly visible on the left edges.
* `global_map.png` 0.5430546319969413. The line here extends outside the spines just a little bit. It should be clipped, but it's visible on the outer grey side.
* `multiple_projections5.png` 0.5250993556778882. This is mostly just stuff around the edges, but there are a lot of subplots so it raises the RMS. For the other lines, I suspect that things were already different, but RMS was under the original 0.5 tolerance before.
* `pcolormesh_global_wrap1.png` 1.2638301573655641. The pcolormesh extends outside the Axes on the left and left/top edges.
* `pcolormesh_global_wrap2.png` 1.6054726285773642. Same as above.
* `pcolormesh_global_wrap3.png` 1.3892948851045908. Same as above, though I think the Robinson map is a bit better instead.
* `pcolormesh_goode_wrap.png` 0.6293873785056563. Basically the same edge changes. I think this one is also improved a bit.
* `nightshade_platecarree.png ` 1.1415358805866185. The inner grey edge along the top is much darker now and it doesn't match the inner grey on the left and right edges.
* `poly_interiors.png` 3.0048004417006537. This is outright a bug in Matplotlib. https://github.com/matplotlib/matplotlib/issues/15946

greglucas commented:
Are you going to close this draft PR and merge all the changes in that other link you sent?
The changed images all look good to me. Minor and likely due to edge antialiasing as you mentioned. The only question I have is in regards to your last bullet.

> poly_interiors.png 3.0048004417006537. This is outright a bug in Matplotlib. matplotlib/matplotlib#15946

Do you want to update this image if it is a known bug, or just xfail it in pytest for now and then update the image once that is fixed in MPL upstream?

greglucas commented:
@QuLogic, are you still planning on doing more work on this? I think it is an excellent improvement and well worth getting in for the next release.

QuLogic commented:
I still want to do this for 0.18, but I'm looking to fix the two bugs in Matplotlib first. I had half-fixed one, and the other should be possible, though it needs further testing.

QuLogic commented:
I've now run through the whole combination of dependencies, and updated the tests to account for the change here. I updated the images where stuff was improved and where tolerances were small-ish, I just increased them (note our default is 0.5 RMS.) Images were generated with the latest Matplotlib, and tolerances added for older versions if necessary.

Some much older versions of Matplotlib are completely broken for some tests, but rather than fix that, I just decided to mark the tests as `xfail`. We're going to drop support for old Matplotlib anyway.

Unfortunately, I was not able to fix some of the above bugs in Matplotlib for 3.2.0, so I just upped the tolerances for the tests that were affected here.

QuLogic commented:
Oh, and one other is `regrid_image.png` is not updated yet. The clipping works better for the first Axes, but the extra clipped pixels on the red/green/blue images is a bug somewhere, and I haven't figured out whether it's Matplotlib or Cartopy yet. We can update that image later.

greglucas commented on lib/cartopy/mpl/slippy_image_artist.py:
```diff
@@ -40,7 +40,7 @@ class SlippyImageArtist(AxesImage):
         self.raster_source = raster_source
         super(SlippyImageArtist, self).__init__(ax, **kwargs)
-        self.set_clip_path(ax.background_patch)
+        self.set_clip_path(ax.patch)
```
Now that the background_patch is in the axes, won't this artist be clipped automatically by your new `ax.patch`?

QuLogic replied:
You're right, I've remove this.

QuLogic commented on lib/cartopy/mpl/geoaxes.py:
```diff
@@ -1416,50 +1431,25 @@ def set_boundary(self, path, transform=None, use_as_clip_path=True):
-            square in axes coordinates.
 
         """
+        if use_as_clip_path is not None:
```
I prefer that people use Matplotlib's builtin support for clipping, by setting `clip_on=False` on the artists they don't want to be clipped.

The `regrid_image.png` test image is affected by this, but honestly it looks better this way (note, not updated yet due to reasons above.)
